core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalReader.java

```diff
@@ -11,101 +11,266 @@
    *
    * @param journal the handle to the journal
    */
-  UfsJournalReader(UfsJournal journal) {
+  UfsJournalReader(UfsJournal journal, JournalReaderOptions options) {
+    mFilesToProcess = new ArrayDeque<>();
     mJournal = Preconditions.checkNotNull(journal, "journal");
-    mUfs = UnderFileSystem.Factory.get(mJournal.getLocation().toString());
-    mCheckpoint = mJournal.getCheckpoint();
+    mNextSequenceNumber = options.getNextSequenceNumber();
+    mPrimary = options.isPrimary();
+    mCheckpointPeriodEntries = Configuration.getLong(
+        PropertyKey.MASTER_JOURNAL_CHECKPOINT_PERIOD_ENTRIES);
   }
 
   @Override
-  public boolean isValid() {
-    return mCheckpointRead && (mCheckpointOpenedTime == mCheckpointLastModifiedTime);
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+    if (mInputStream != null) {
+      mInputStream.close();
+    }
   }
 
   @Override
-  public JournalInputStream getCheckpointInputStream() throws IOException {
-    if (mCheckpointRead) {
-      throw new IOException("Checkpoint file has already been read.");
+  public boolean shouldCheckpoint() throws IOException {
+    if (mPrimary) {
+      return false;
     }
-    mCheckpointOpenedTime = getCheckpointLastModifiedTimeMs();
-
-    LOG.info("Opening journal checkpoint file: {}", mCheckpoint);
-    JournalInputStream jis =
-        mJournal.getJournalFormatter().deserialize(mUfs.open(mCheckpoint.toString()));
+    if (mNextSequenceNumber > UfsJournalSnapshot.getNextLogSequenceToCheckpoint(mJournal)
+        && mNextSequenceNumber % mCheckpointPeriodEntries == 0) {
+      return true;
+    }
+    return false;
+  }
 
-    mCheckpointRead = true;
-    return jis;
+  @Override
+  public long getNextSequenceNumber() {
+    return mNextSequenceNumber;
   }
 
   @Override
-  public JournalInputStream getNextInputStream() throws IOException {
-    if (!mCheckpointRead) {
-      throw new IOException("Must read the checkpoint file before getting input stream.");
+  public Journal.JournalEntry read() throws IOException, InvalidJournalEntryException {
+    while (true) {
+      Journal.JournalEntry entry = readInternal();
+      if (entry == null) {
+        return null;
+      }
+      if (mInputStream.mFile.isCheckpoint()) {
+        return entry;
+      }
+      if (entry.getSequenceNumber() == mNextSequenceNumber) {
+        mNextSequenceNumber++;
+        return entry;
+      }
+      if (entry.getSequenceNumber() < mNextSequenceNumber) {
+        // This can happen in the following two scenarios:
+        // 1. The primary master failed when renaming the current log to completed log which might
+        //    result in duplicate logs.
+        // 2. The first completed log after the checkpoint's last sequence number might contains
+        //    some duplicate entries with the checkpoint.
+        LOG.debug("Skipping duplicate log entry {}.", entry);
```

Comment:
Could this be more detailed, and include the current SN and the entry SN (might already be in the entry output).